hedera-node/src/main/java/com/hedera/services/fees/charging/NonHapiFeeCharging.java

```diff
@@ -0,0 +1,92 @@
+ */
+package com.hedera.services.fees.charging;
+
+import static com.hedera.services.exceptions.ValidationUtils.validateResourceLimit;
+import static com.hedera.services.ledger.properties.AccountProperty.BALANCE;
+import static com.hedera.services.ledger.properties.AccountProperty.IS_DELETED;
+import static com.hedera.services.state.submerkle.EntityId.MISSING_ENTITY_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_BALANCES_FOR_STORAGE_RENT;
+
+import com.hedera.services.ledger.TransactionalLedger;
+import com.hedera.services.ledger.properties.AccountProperty;
+import com.hedera.services.state.merkle.MerkleAccount;
+import com.hedera.services.state.submerkle.EntityId;
+import com.hederahashgraph.api.proto.java.AccountID;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * Responsible for charging the auto-renewal fee and storage rent from the payer or auto-renew
+ * account if present on contracts and accounts. Distributes the charged fee between 0.0.800,
+ * 0.0.801 and the funding account
+ */
+public class NonHapiFeeCharging {
+    // Used to distribute charged fee to collection accounts in correct percentages
+    final FeeDistribution feeDistribution;
+
+    @Inject
+    public NonHapiFeeCharging(final FeeDistribution feeDistribution) {
+        this.feeDistribution = feeDistribution;
+    }
+
+    public void chargeNonHapiFee(
+            @Nullable final EntityId preferredPayer,
+            final AccountID finalPayer,
+            final long amount,
+            TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accounts) {
+        chargeFee(preferredPayer, finalPayer, amount, accounts);
+    }
+
+    private void chargeFee(
+            @Nullable final EntityId preferredPayer,
+            final AccountID finalPayer,
+            final long fee,
+            final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accounts) {
+        var leftToPay = fee;
+
+        if (preferredPayer != null && !MISSING_ENTITY_ID.equals(preferredPayer)) {
+            final var grpcId = preferredPayer.toGrpcAccountId();
+            if (accounts.contains(grpcId) && !(boolean) accounts.get(grpcId, IS_DELETED)) {
+                final var debited =
+                        charge(preferredPayer.toGrpcAccountId(), leftToPay, false, accounts);
+                leftToPay -= debited;
+            }
+        }
+        if (leftToPay > 0) {
+            charge(finalPayer, leftToPay, true, accounts);
+        }
+    }
+
+    private long charge(
+            final AccountID payer,
+            final long amount,
+            final boolean isLastResort,
+            final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accounts) {
+        long paid;
+        final var balance = (long) accounts.get(payer, BALANCE);
+        if (amount > balance) {
+            validateResourceLimit(!isLastResort, INSUFFICIENT_BALANCES_FOR_STORAGE_RENT);
```

Comment:
We probably want an extra `ResponseCodeEnum failureStatus` argument passed in for this check, so we can use different statuses when charging storage rent vs. charging auto-renewal fees.

Reply:
Agreed. Do you think we need a new Response Code `INSUFFICIENT_BALANCES_FOR_AUTORENEWAL` or use `INSUFFICIENT_PAYER_BALANCE`